Source/Applications/ApplicationStructureFragment.cs

```diff
@@ -168,5 +172,65 @@ void ThrowIfAnyChildDoesNotBelongToType(IEnumerable<IApplicationStructureFragmen
         {
             children.ForEach(fragment => ThrowIfTypeDoesNotBelongToParent(fragment.Type, type));
         }
+
+        /// <inheritdoc/>
+        public bool Equals(IApplicationStructureFragment other)
+        {
+            if (Type != other.Type) return false;
+            if (Recursive != other.Recursive) return false;
+            if (Required != other.Required) return false;
+            if (HasParent != other.HasParent) return false;
+
+            if (HasParent == true && other.HasParent == true) 
+            {
+                if (! Parent.Equals(other.Parent)) return false;
```

Comment:
Hmm, we might now want to do this. This smells like infinite recursion.
Suggestions @einari ?

Reply:
The parent is just an representation of the Parent IApplicationStructureFragment's state at the time this ApplicationStructureFragment is constructed, therefor I don't think that the Parents can mean anything when doing comparisons.

Maybe just remove checks on Parent from Equals and GetHashCode?
@einar